lib/plugins/plugin/plugin.js

```diff
@@ -0,0 +1,343 @@
+
+    this.hooks = {
+      'plugin:install:install': () => BbPromise.bind(this)
+        .then(this.install)
+        .then(this.trackPluginInstall),
+      'plugin:uninstall:uninstall': () => BbPromise.bind(this)
+        .then(this.uninstall)
+        .then(this.trackPluginUninstall),
+      'plugin:list:list': () => BbPromise.bind(this)
+        .then(this.list)
+        .then(this.trackPluginList),
+      'plugin:search:search': () => BbPromise.bind(this)
+        .then(this.search)
+        .then(this.trackPluginSearch),
+    };
+  }
+
+  install() {
+    return BbPromise.bind(this)
+      .then(this.validate)
+      .then(this.getPlugins)
+      .then((plugins) => {
+        const pluginName = this.options.name;
+        const plugin = plugins.find((item) => item.name === pluginName);
+
+        if (plugin) {
+          const servicePath = this.serverless.config.servicePath;
+          const packageJsonFilePath = path.join(servicePath, 'package.json');
+
+          // check if package.json is already present. Otherwise create one
+          if (!fs.existsSync(packageJsonFilePath)) {
+            this.serverless.cli
+              .log('Creating an empty package.json file in your service directory');
+
+            const packageJsonFileContent = {
+              name: this.serverless.service.service,
+              description: '',
+              version: '0.1.0',
+              dependencies: {},
+              devDependencies: {},
+            };
+
+            fse.writeJsonSync(packageJsonFilePath, packageJsonFileContent);
+          }
+
+          // install the package through npm
+          this.serverless.cli
+            .log(`Installing plugin "${pluginName}" (this might take a few seconds...)`);
+
+          childProcess
+            .execSync(`npm install --save-dev ${pluginName}`, {
+              stdio: 'ignore',
+            });
+
+          // check if plugin was installed correctly
+          const pluginInstalled = !!JSON.parse(
+            fs.readFileSync(packageJsonFilePath).toString()
+          ).devDependencies[pluginName];
+
+          if (pluginInstalled) {
+            const serverlessFilePath = this.getServerlessFilePath();
+            let serverlessFileContent = fs.readFileSync(serverlessFilePath).toString();
+
+            const newPluginsArray = {
+              plugins: [],
+            };
+
+            // load the service and parse it to JSON
+            const parsedYaml = YAML.load(serverlessFileContent);
+
+            if (parsedYaml.plugins) {
+              newPluginsArray.plugins = parsedYaml.plugins;
+            }
+
+            newPluginsArray.plugins = _.union(newPluginsArray.plugins, [pluginName]);
+
+            if (parsedYaml.plugins) {
+              // replace the plugins definition in the serverless file
+              serverlessFileContent = serverlessFileContent
+                .replace(
+                  /(plugins:)+(\s)*(- .*\s*)*(\[\])*/, // eslint-disable-line no-useless-escape
```

Comment:
@horike37 we need to use a YAML ast parser. The framework in general needs to use an AST parser to avoid regex issues.

see https://github.com/serverless/dashboard/blob/master/package.json#L125

& https://github.com/serverless/dashboard/blob/master/app/utils/yaml/updateYamlContents.js

Lets get the YAML AST parse into framework core and as a /util function that anything can use